### PR TITLE
[Closed]

### LINKED ISSUE #9811: [BOOM-4293] Delete the main category of a product, product should be reassigned another main category
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-4293).

- _**Reporter:**_ marion_francois
- _**Created at:**_ Mon, 13 Nov 2017 11:21:40 +0100

#### Describe the bug

When you associated a product to two categories and delete the default one with the option **If they have no other category, I want to associate them with the parent category.**. My product doesn't have a default category (NOK!).

#### Expected behavior

It should have the parent category as default category.

**TO SUM IT UP**: When I delete the main category of product A, then I am prompted a choice. If I choose that product A has to be re-associated with another category B, then category B must become the "default category" for product A.

#### Steps to reproduce

1. I associate my product to 2 categories: **Men** & **Home**
2. The default category for my products is **Men**
3. Delete the category **Men**
4. Choose to associate **the products without other categories** to the parent category, and keep them enabled.
5. My product is still associated to the Home category (good !) but doesn't have a default category (bad!).
       
**Screenshots**

https://drive.google.com/file/d/1BRyAAbhssGSG_Co8Y1GOZdYcNPYKMam1/view?usp=sharing

capture d'écran_192

#### Additional infos

- PrestaShop version: 1.7.7.3
- PHP version: 7.2

### PR DESCRIPTION
**Pull request by arouiadib**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | Closed as it is a BC break, see second approach PR
| Type?             | bug fix
| Category?         | BO 
| BC breaks?        | yes
| Deprecations?     |  no
| Fixed ticket?     | Fixes #9811
| How to test?      | Closed, see discussion
| Possible impacts? | Closed, see discussion

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24223)

### CODE REVIEW

File: admin-dev/themes/default/js/bundle/product/default-category.js
Comment (arouiadib) [inline admin-dev/themes/default/js/bundle/product/default-category.js:4]: This modif should be removed from this PR before merge. (It just help testing this PR easily and seperately)

File: classes/Category.php
Comment (sowbiba) [inline classes/Category.php]: I'm wondering about category deletions that doesn't use the command handlers. I'm not very familiar with the webservice but does it use commands ? @PrestaShop/prestashop-maintainers 
Just calling category->delete() will lead to DB inconsistency. How can we be sure that won't happen ?

**Review by sowbiba — CHANGES_REQUESTED** [review]

**Review by matks — COMMENTED** [review]
Hello @arouiadib I dont understand what is the logic behind the PR Can you explain?

I dont understand why you add some delete logic? I think https://github.com/PrestaShop/PrestaShop/issues/9811 can be fixed if we add some "re-assign main category ID" logic ?

**Comment by arouiadib** [comment]
Hi @sowbiba , @matks and thanks a lot for the time reviewing my PR.

I don't add any deletion logic in this PR.

There were a function named `cleanAssoProducts` which I only moved from legacy model to new deletion handlers. This method should happen after and not before "re-assign main category ID".

`cleanAssoProducts` doesn't delete any product, and doesn't delete any category, it only cleans the associations (the rows in join table `ps_category_product`). The problem is in the comment of the signature of this method, which is misleading IMO ( should be `remove associations` instead of `remove associated products`).

What I did is move this function at the end of the process. Why? Because in deletion handler there is a SQL query that needs those associations to exist in DB. That query return no results if `cleanAssoProducts` is already executed. 

I also think there was a problem in SQL query itself so I tried to fix it.

@matks, the "re-assign main category ID" logic was always there and is correctly working. The problem was in SQL and in the order of execution.

I know this matter is bit complex and not easy to explain, but believe me it wasn't  a few-minutes PR. I will try to enhance it ASAP by some tests if I can.

@sowbiba, I think I got your point and still investigate if deletion of category via webservice follow different logic.

**Comment by matks** [comment]
> There were a function named `cleanAssoProducts` which I only moved from legacy model to new deletion handlers. This method should happen after and not before "re-assign main category ID".

OK, thank you for explanations. Explaining [your intent](https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests) is sometimes as important as the PR itself.

> `cleanAssoProducts` doesn't delete any product, and doesn't delete any category, it only cleans the associations (the rows in join table `ps_category_product`). The problem is in the comment of the signature of this method, which is misleading IMO ( should be `remove associations` instead of `remove associated products`).

Indeed the naming is ambiguous 😆 this is why I was confused.

> What I did is move this function at the end of the process. Why? Because in deletion handler there is a SQL query that needs associations to exist in DB. That query return no results if `cleanAssoProducts` is already executed.

I am afraid of the side-effects of moving the time where `cleanAssoProducts()` is called, in other sections of PrestaShop. Maybe instead of doing this we can improve how `the SQL query that needs associations to exist in DB` works? We could extract the associations data before it is deleted and use it later.

> @matks, the "re-assign main category ID" logic was always there and is correctly working.

I dont see where the "re-assign main category ID" , can you help me find it ?

> I know this matter is bit complex and not easy to explain, but believe me it wasn't a few-minutes PR. I will try to enhance it ASAP by some tests if I can.

I think some Behat tests like https://github.com/PrestaShop/PrestaShop/blob/develop/tests/Integration/Behaviour/Features/Scenario/Category/category_management.feature would be great

```
  Scenario: Delete category and verify products have been reassigned
    When I add new category "categoryA" with following details:
      ...
    And "productB" is associated with "categoryA"
    And "productB" main category is "categoryA"
    And I delete categories "categoryA" choosing mode "associate_and_disable"
    Then category "category1" does not exist
    And "productB" main category is "categoryC" (category C is category A parent)
```

**Comment by arouiadib** [comment]
Hi @matks , thanks for the time, detailed answer, and usual help.

I understand now why it is a possible BC. :smiley: 

Extracting the association data, before category is deleted (to use it later) seems a good idea. I will give it more thought and see if I can create a (maintainable) solution with this approach. 

The "re-assign main category ID" logic is in `PrestaShop\PrestaShop\Adapter\Category\CommandHandler\AbstractDeleteCategoryHandler`  [Line 69](https://github.com/PrestaShop/PrestaShop/blob/5d615aa6e4fca8a80929223aade271237d98c62a/src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php#L69) . It happens via a loop through products returned by SQL.

Thanks for the sample of Behat test, sure it will help me write test for this case.

Again thanks.

File: classes/Category.php
Comment (arouiadib) [reply classes/Category.php]: Very nice catch.

**Comment by arouiadib** [comment]
Hi @matks , could you please see the PR created in favor of your approach? #24346

**Pull request closed without merging**
